gadget/raw.go

```diff
@@ -0,0 +1,281 @@
+// or determined to be identical will not be analyzed on subsequent calls.
+func (r *RawStructureUpdater) Backup(from *PositionedStructure, to *PositionedStructure) error {
+	device, err := r.findDevice(to)
+	if err != nil {
+		return fmt.Errorf("cannot find device matching structure %v: %v", to, err)
+	}
+	disk, err := os.OpenFile(device, os.O_RDONLY, 0)
+	if err != nil {
+		return fmt.Errorf("cannot open device for reading: %v", err)
+	}
+	defer disk.Close()
+
+	for idx, pc := range to.PositionedContent {
+		if err := r.backupOrCheckpointContent(disk, &pc, r.contentBackupPath(to, &pc, idx)); err != nil {
+			return fmt.Errorf("cannot backup image %v: %v", pc, err)
+		}
+	}
+
+	return nil
+}
+
+func (r *RawStructureUpdater) rollbackDifferent(out io.WriteSeeker, pc *PositionedContent, rw *RawStructureWriter, backupPath string) error {
+	if osutil.FileExists(backupPath + ".same") {
+		// content the same, no update needed
+		return nil
+	}
+
+	backup, err := os.Open(backupPath + ".backup")
+	if err != nil {
+		return fmt.Errorf("cannot open backup image: %v", err)
+	}
+
+	if err := rw.WriteRawStream(out, pc, backup); err != nil {
+		return fmt.Errorf("cannot restore backup: %v", err)
+	}
+
+	return nil
+}
+
+// Rollback attempts to restore original content from backup copies prepared during Backup().
+func (r *RawStructureUpdater) Rollback(from *PositionedStructure, to *PositionedStructure) error {
+	device, err := r.findDevice(to)
+	if err != nil {
+		return fmt.Errorf("cannot find device matching structure %v: %v", to, err)
+	}
+
+	disk, err := os.OpenFile(device, os.O_WRONLY, 0)
+	if err != nil {
+		return fmt.Errorf("cannot open device for writing: %v", err)
+	}
+	defer disk.Close()
+
+	rw := NewRawStructureWriter(r.rootDir)
+
+	for idx, pc := range to.PositionedContent {
+		if err := r.rollbackDifferent(disk, &pc, rw, r.contentBackupPath(to, &pc, idx)); err != nil {
+			return fmt.Errorf("cannot rollback image %v: %v", pc, err)
+		}
+	}
+
+	return nil
+}
+
+func (r *RawStructureUpdater) updateDifferent(disk io.WriteSeeker, pc *PositionedContent, rw *RawStructureWriter, backupPath string) error {
+	if osutil.FileExists(backupPath + ".same") {
+		// content the same, no update needed
+		return nil
+	}
+
+	if !osutil.FileExists(backupPath + ".backup") {
+		// not the same, but a backup file is missing, error out just in
+		// case
+		return fmt.Errorf("missing backup file")
+	}
+
+	if err := rw.WriteRawImage(disk, pc); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Update attempts to update given structure. The structure must have been
+// analyzed and backed up by a prior Backup() call.
+func (r *RawStructureUpdater) Update(from *PositionedStructure, to *PositionedStructure) error {
+	device, err := r.findDevice(to)
+	if err != nil {
+		return fmt.Errorf("cannot find device matching structure %v: %v", to, err)
+	}
+
+	disk, err := os.OpenFile(device, os.O_WRONLY, 0)
```

Comment:
silly question, we expect the relevant devices to be not busy, right?

Reply:
Yes. Also, I don't think we can try the special case of O_EXCL for block devices. IIRC it fails with EBUSY if a partition from that device is mounted, which makes sense.